eval/builtin_special.go

```diff
@@ -447,27 +447,41 @@ func compileWhile(cp *compiler, fn *parse.Form) effectOpBody {
 	args := cp.walkArgs(fn)
 	condNode := args.next()
 	bodyNode := args.nextMustLambda()
+	elseNode := args.nextMustLambdaIfAfter("else")
 	args.mustEnd()
 
-	return &whileOp{cp.compoundOp(condNode), cp.primaryOp(bodyNode)}
+	condOp := cp.compoundOp(condNode)
+	bodyOp := cp.primaryOp(bodyNode)
+	var elseOp valuesOp
+	if elseNode != nil {
+		elseOp = cp.primaryOp(elseNode)
+	}
+
+	return &whileOp{condOp, bodyOp, elseOp}
 }
 
 type whileOp struct {
-	condOp, bodyOp valuesOp
+	condOp, bodyOp, elseOp valuesOp
 }
 
 func (op *whileOp) invoke(fm *Frame) error {
 	body := op.bodyOp.execlambdaOp(fm)
+	elseFn := op.elseOp.execlambdaOp(fm)
+	isFirstLoop := true
```

Comment:
Can you use the same pattern as in the implementation of `for`, keeping track of whether any iteration has been executed at all, and execute the `else` body if no iteration has been executed?